include/multi_blas_helper.cuh

```diff
@@ -121,32 +194,32 @@ namespace quda
+    template <typename Functor>
+    inline int max_YW_size(int NXZ, QudaPrecision x_prec, QudaPrecision y_prec)
     {
       bool x_fixed = x_prec < QUDA_SINGLE_PRECISION;
       bool y_fixed = y_prec < QUDA_SINGLE_PRECISION;
-      size_t scalar_size = scalar_width * std::max(std::max(x_prec, y_prec), QUDA_SINGLE_PRECISION);
-      NXZ = is_valid_NXZ(NXZ, reduce, x_fixed) ? NXZ : MAX_MULTI_BLAS_N; // ensure NXZ is a valid size
+      NXZ = is_valid_NXZ(NXZ, Functor::reducer, x_fixed) ? NXZ : MAX_MULTI_BLAS_N; // ensure NXZ is a valid size
       size_t spinor_x_size = x_fixed ? sizeof(Spinor<short, 4>) : sizeof(Spinor<float, 4>);
       size_t spinor_y_size = y_fixed ? sizeof(Spinor<short, 4>) : sizeof(Spinor<float, 4>);
-
       size_t spinor_z_size = spinor_x_size;
       size_t spinor_w_size = x_fixed ? sizeof(Spinor<short, 4>) : sizeof(Spinor<float, 4>);
 
       // compute the size remaining for the Y and W accessors
-      int arg_size = (MAX_ARG_SIZE - sizeof(int)                       // NYW parameter
-                      - NXZ * spinor_x_size                            // SpinorX array
-                      - (use_z ? NXZ * spinor_z_size : sizeof(void *)) // SpinorZ array (else dummy pointer)
-                      - 2 * sizeof(int)                                // functor NXZ/NYW members
-                      - (multi_1d ? scalar_size * 3 * max_N_multi_1d() : 0) // multi_1d coefficient arrays
-                      - sizeof(int)                                    // length parameter
-                      - (!use_w ? sizeof(void *) : 0)                  // subtract dummy pointer if not using W
-                      - (reduce ? sizeof(ReduceArg<device_reduce_t>) : 0)        // reduction buffers
-                      - 16) // there seems to be 16 bytes other argument space we need
-        / (spinor_y_size + (use_w ? spinor_w_size : 0));
+      const auto arg_size = (max_arg_size<Functor>()
+                             - sizeof(kernel_param<>)                                      // kernel_param parent
+                             - sizeof(int)                                                 // NYW parameter
+                             - NXZ * spinor_x_size                                         // SpinorX array
+                             - (Functor::use_z ? NXZ * spinor_z_size : sizeof(void *))     // SpinorZ array (else dummy pointer)
+                             - sizeof(Functor)                                             // functor
+                             - sizeof(dim3)                                                // threads parameter
+                             - (!Functor::use_w ? sizeof(void *) : 0)                      // subtract dummy pointer if not using W
+                             - (Functor::reducer ? sizeof(ReduceArg<device_reduce_t>) : 0) // reduction buffers
+                             - 12) // extra 12 bytes subtracted for alignment roundup
```

Comment:
Same comment about magic numbers above

Reply:
Resolved: this factor of 12 is no longer needed.